### DISCUSSION TITLE
Change the color of an object

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1227] [General]
I'm exploring how to Vulkan Scene Graph in a CAD context and as part of that I'd like to be able to click on an object and make it "selected", and make it show visually. 

To make things simpler, I'd initially like to change the color of a box created by the Builder. By dumping the box created by the builder as a vsgt file by `vsg::write()`, I found that the color is resides in a `vsg4Array` of size 1 with id=6. But how would I find this node programmatically, and how would I set its value and refresh the view?

I'd also like to be able to do the same with an STL file loaded by `vsg::read_cast<vsg::Node>()`, and I wonder how would I change the color of such a node. However I'm wondering if `vsgXchange` does not add a lot of nodes and memory model that would be "redundant" in a CAD context?

**Comment by jacky-wang7** [comment]
Find the color vsg4Array node. This example may help you：“vsgExamples: state\vsgdynamicvertex ”.

**Reply by dov** [reply]
Thanks! Indeed by using a visitor I was able to find the color node, and change its color. Here is code that shows how to find and change the color of a builder generated object:

```c++
    // Set box color to red
    geomInfo.color = vsg::vec4 {1.0f, 0.0f, 0.0f, 1.0f};
    auto node = builder->createBox(geomInfo, stateInfo);

    // Find the color node and change its color
    struct ChangeColor : public vsg::Visitor
    {
        void apply(vsg::Object& object) override
        {
            object.traverse(*this);
        }
    
        void apply(vsg::VertexIndexDraw& vid) override
        {
            for (auto& buffer : vid.arrays) {
                auto data = buffer->data;
                if (data->is_compatible(typeid(vsg::vec4Array))) {
                    // Change the color to green
                    (*(vsg::vec4*)(data->dataPointer(0)))
                        = vsg::vec4 {0.0, 1.0, 0.0, 1.0};
                }
            }
        }
    };
    
    ChangeColor fc;
    node->accept(fc);
```

**Reply by dov** [reply]
In my example above, I changed the color before the pipeline was created and then the color change applied. I now wanted to use this behavior in order to change the color of an object that I select by using the LineIntersector, i.e. after the pipeline was prepared. Even though I succesfully was able to locate the node and apply the color change, did the actual color not change in the viewer. I assume I have to tell the vsg to refresh the scene. But how do I do that?

I did call `data->dirty()` after changing the color, but apparently that was not enough.

**Reply by jacky-wang7** [reply]
Maybe you can check if color->properties.dataVariance is vsg::DYNAMIC_DATA.
DYNAMIC_DATA and STATIC_DATA have different update methods. 
They are demonstrated separately in vsgdynamicvertex.

**Reply by dov** [reply]
Thanks! This was indeed the problem! Once I changed the dataVariance to vsg::DYNAMIC_DATA it worked!

Below is a link to my example of how to do picking and change the color of the selected object. I'd be happy to receive comments of how to improve it.

Also let me know if there is any interest that I submit it as a vsgExample.

See:

https://gist.github.com/dov/11d43a92db9eaae5e1d55043053474ea

**Comment by robertosfield** [comment]
You can dive into the subgraph and modifying the existing colour as suggested above or add your own behaviors, like additional colours used for picking, you can modifying the vsg::ShaderSet that the loaders use by setting the vsg::Options::shaderSets[] or by setting the vsg::Builder.shaderSet.

The vsg::ShaderSet provides a way of specifying what uniforms, textures and and vertex arrays should be used with a set of shaders, along with #define support for enabling/disabling different code paths in the shaders.

**Reply by dov** [reply]
Thanks for your answer @robertosfield . However, even after looking at the sources and examples, I fail to understand how the shaderSets are used, and how they can be used to push a color change. Perhaps I'll just wait until you finish the tutorial.

**Reply by robertosfield** [reply]
I have been so swamped with development work I haven't been able to do further work on tutorial in the past 6 months, alas I will continue this busy through the summer so please don't wait on my completing the tutorial - writing it takes a lot of work so not easy to just do when my other work is so full on.

The idea behind ShaderSet is that it allows a level of decoupling between the creating of scene graphs and the state and associated shaders that it is built with.  Without ShaderSet you wouldn't have able way to customize how things are rendered when you load them from vsgXchange and have them built with vsg::Builder etc.

In your particular case the way I would implement it would be to have uniform that is used to set the picked colour/status and any other object wide behaviors that you want.  This uniform could be packed into a struct like the material structs in include/vsg/state/material.h[](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/state/material.h).  These would be encapsulated with a vg::Value<> and then assigned to a StateGroup->BindDescritptorSet->DescritporSet->DescriptorBuffer->BufferInfo->data, then you place the StateGroup above the subgraph you want to highlight.
